ELEC-259: Proposal to support clang-format

ckitagawa opened the pull request:
TL;DR: Added clang-format it caused a few common changes I couldn't figure out how to avoid. Let's discuss if we are OK with them. If not how do we want to move forward.

[Clang-Format Documentation](https://clang.llvm.org/docs/ClangFormatStyleOptions.html) for those interested. 

I'm fine with people pulling, modifying and committing the `.clang-format` file in this branch if they want to try to avoid any of these changes.

Added a `.clang-format` file to root directory. It is based on the Google style clang-format with modifications that support some of our coding conventions as outlined on Confluence. Specifically, it modifies:

* ColumnLimit: 100
* PointerAlignment: Right
* Cpp11BracedListStyle: false

**Major deviations from current style**
```c
// Empty Loops
while (cond) { }
// becomes
while (cond) {
}

// Space in Empty Braces
void func(void) { }
// becomes
void func(void) {}

// Designated Struct Initializers
typedef struct TestStruct {
  int test1;
  int test2;
} TestStruct;

const static TestStruct ts = {
  .test1 = 1
};
// becomes
const static TestStruct ts = {.test1 = 1 };  
// Note: that the {. without a space is fixed by an LLVM patch which has yet to be released: 
// https://reviews.llvm.org/rL305696#9b94b34f
// This can be fixed with a trailing comma. 
const static TestStruct ts = {
  .test1 = 1,
};
// but poses a different problem as
const static TestStruct ts = {
  .test1 = 1,
  .test2 = 2,
};
// becomes
const static TestStruct ts = {
  .test1 = 1, .test2 = 2,
};
// If the designated initializers manage to fit on one line.
```

**Common changes not mentioned in the style guide**

* Strictly orders header includes to be: corresponding header if a `*.c` file, `<*.h>`, `"*.h"`. This is a feature I wanted anyway :p
* BinPacking and argument alignment in function definitions. (I don't think we should care since this doesn't have significant effect).
* Right aligning the multiline character `\`. (Mainly in macros).
* Indentation in macros starting at 2 spaces instead of none
```c
#define MY_MACRO(x)                                                            \
do_something((x))
\\ becomes
#define MY_MACRO(x)                                                            \
  do_something((x))
```

**Summary**

I'm personally in favor of this change to using `.clang-format` as it just makes everything simpler and more consistent. No more discussion of style and since the linter already supports most of these changes we can save time by using it. Moreover, clang-format has built-in support in most common editors. 

Personally, I am fine with the few minor changes this causes but I know others may not be. Let's discuss concerns here and work towards a way to use this tool or another like it to simplify reviews by being able to trust the style is being followed.

@tituschow I've tried the `BreakBeforeBraces: Custom` and `BraceWrapping: AfterStruct: true` but it doesn't do what you would think:
```c
typedef struct hello {
  int hi;
} hello;
\\ becomes
typedef struct hello
{
  int hi;
} hello;
```

ckitagawa commented:
Not sure what's up with the lint errors... It has to do with @Derpalooza 's method for casting initializer stucts in `digital_io.c`. I'll look into it but I may just end up rewriting that bit...

tituschow commented on libraries/ms-common/src/can_rx.c:
```diff
@@ -40,20 +40,19 @@ StatusCode can_rx_register_handler(CANRxHandlers *rx_handlers, CANMessageID msg_
-    .context = context
-  };
+  rx_handlers->storage[rx_handlers->num_handlers++] =
+      (CANRxHandler){.msg_id = msg_id, .callback = handler, .context = context };
```
Was this supposed to become one line?

tituschow commented on libraries/ms-common/src/can.c:
```diff
@@ -1,15 +1,13 @@
-    .source_id = can->device_id,
-    .type = msg->type,
-    .msg_id = msg->msg_id
+    .source_id = can->device_id, .type = msg->type, .msg_id = msg->msg_id,
```
Single line?

tituschow commented on libraries/ms-common/src/can_rx.c:
```diff
@@ -40,20 +40,19 @@ StatusCode can_rx_register_handler(CANRxHandlers *rx_handlers, CANMessageID msg_
 CANRxHandler *can_rx_get_handler(CANRxHandlers *rx_handlers, CANMessageID msg_id) {
-  const CANRxHandler key = { .msg_id = msg_id };
+  const CANRxHandler key = {
+    .msg_id = msg_id,
```
Single line?

tituschow commented on libraries/ms-common/src/stm32f0xx/soft_timer.c:
```diff
@@ -113,10 +113,10 @@ static void prv_init_periph(void) {
 
   TIM_TimeBaseInitTypeDef timer_init = {
-    .TIM_Prescaler = (clocks.PCLK_Frequency / 1000000) - 1, // 1 Mhz
+    .TIM_Prescaler = (clocks.PCLK_Frequency / 1000000) - 1,  // 1 Mhz
```
Why does clang format always add at least 2 spaces for comments?

ckitagawa replied:
That's the Google style for inline comments. I don't think it matters?

tituschow replied:
Nah, I was just curious.

tituschow commented:
Yeah, dealing with initializers is kinda weird. Is there no way to keep the current style or at least add a new line for each member?

tituschow commented on libraries/ms-common/src/stm32f0xx/spi.c:
```diff
@@ -11,14 +11,10 @@ typedef struct {
-    .rcc_cmd = RCC_APB2PeriphClockCmd,
-    .periph = RCC_APB2Periph_SPI1,
-    .base = SPI1
+      .rcc_cmd = RCC_APB2PeriphClockCmd, .periph = RCC_APB2Periph_SPI1, .base = SPI1,
```
Was this supposed to be one line? It's kinda hard to figure out what the rule for initializers is. Doesn't seem to be consistent even though you're ending it with a comma?

ckitagawa replied:
So the comma only makes the following change
```c
{ .hello = 1, .test = 1 };
// vs
{
  .hello = 1, .test = 1,
};
```
The multi-line initializer
{
  .hello = 1,
  .test = 1,
};
```
will only happen if the arguments/parameters to the initializer exceed our column limit of 100 characters.

tituschow replied:
Ah, I see. So if we don't have a comma, will it pack the initializer like a function call or multi-line?

ckitagawa replied:
Without the comma it will pack like a function call.

tituschow commented on libraries/ms-common/src/stm32f0xx/uart.c:
```diff
@@ -19,32 +19,22 @@ typedef struct {
-    .base = USART4
-  }
-};
+static UARTPortData s_port[] = { {.rcc_cmd = RCC_APB2PeriphClockCmd,
```
Like compare this to `spi.c`

tituschow commented on libraries/ms-common/src/x86/gpio.c:
```diff
@@ -29,10 +29,12 @@ static uint32_t prv_get_index(const GPIOAddress *address) {
-                                    .resistor = GPIO_RES_NONE,
-                                    .alt_function = GPIO_ALTFN_NONE };
+  GPIOSettings default_settings = {
+    .direction = GPIO_DIR_IN,
```
If we always had this I'd be pretty happy :P

ckitagawa replied:
This isn't possible so far as I can tell, the rule is one per line only if it doesn't fit on one line otherwise put it on one line. Seems to be fixed with that rule and I tried my best to avoid it without any luck.

tituschow replied:
Hmm, is there a specific reason this one turned out nicely?

ckitagawa replied:
There are too many arguments to put this on one line so it did one per line as per the rule. Basically, if it can pack the initializer arguments into a single line it will. If it can't one per line. 

There is a fairly thorough discussion of the  problem and its solutions [here](https://stackoverflow.com/questions/33656800/clang-format-line-breaks).

TL;DR: Solution are
* Use `\\ clang-format [off | on]` comments.
* Add `\\` after each parameter to force a line break.
* Set `ColumnLimit: 0` which disables wrapping but won't mess with initializers (although you must manually format them the first time.

I'm not a fan of any of those solutions and would rather go with its present behavior. IMO it is consistent once you understand the too and while the one line is ever so slightly less readable I don't see significant value in not using the tool for this single problem.

tituschow replied:
Interesting. I think it would be helpful to have this link and a brief explanation of the problem in the clang format file so people know why we have specific options set.

I think using this and using `//` to force line breaks manually is probably the best thing to do right now. Ultimately, this is just a tool to help make our lives easier and we can always change it if we need to. :P

ckitagawa replied:
Done (everywhere).

tituschow commented on libraries/ms-common/test/test_adc.c:
```diff
@@ -151,9 +153,7 @@ void test_read_continuous() {
-    { .port = GPIO_PORT_A },
-    { .port = GPIO_PORT_B },
-    { .port = GPIO_PORT_C }
+    {.port = GPIO_PORT_A }, {.port = GPIO_PORT_B }, {.port = GPIO_PORT_C }
```
I'm going to stop pointing out every instance it looks weird now

tituschow commented:
Overall, I'm a fan of using tools to enforce our style instead of relying on code review - it looks pretty decent so far! The only thing that's consistently inconsistent are initializers, which is too bad.

ckitagawa commented:
I've gone and fixed what I think are all the blocking issues with initializer formatting and added documentation in the `.clang-format` file. 

If I missed any cases with initializers just comment them and I'll get them fixed. Otherwise this should be everything if we are OK with the other changes.

ckitagawa commented on libraries/ms-common/src/stm32f0xx/i2c.c:
```diff
@@ -1,46 +1,46 @@
-  [I2C_PORT_1] = { .periph = RCC_APB1Periph_I2C1, .base = I2C1 },
-  [I2C_PORT_2] = { .periph = RCC_APB1Periph_I2C2, .base = I2C2 }
-};
+static I2CPortData s_port[NUM_I2C_PORTS] = {[I2C_PORT_1] = {.periph = RCC_APB1Periph_I2C1,
```
The spacing of `{.` on this line is a bug in clang-format that was fixed in late July it should be fixed with the next release.

tituschow commented on libraries/ms-common/src/stm32f0xx/spi.c:
```diff
@@ -9,14 +9,16 @@ typedef struct {
-      .rcc_cmd = RCC_APB1PeriphClockCmd, .periph = RCC_APB1Periph_SPI2, .base = SPI2,
-  }
-};
+static SPIPortData s_port[SPI_MCU_NUM_PORTS] = { {
```
Can you add `[SPI_PORT_x]`?

ckitagawa replied:
Done.

tituschow commented on libraries/ms-common/src/stm32f0xx/spi.c:
```diff
@@ -9,16 +10,12 @@ typedef struct {
-                                                     .periph = RCC_APB1Periph_SPI2,      //
-                                                     .base = SPI2,                       //
-                                                 } };
+static SPIPortData s_port[SPI_MCU_NUM_PORTS] = {[SPI_PORT_1] = {.rcc_cmd = RCC_APB2PeriphClockCmd,
```
Ahhh I see now
